crates/core/src/nats/client.rs

```diff
@@ -0,0 +1,262 @@
+    ) -> Result<NatsConsumer<PullConsumerConfig>, NatsError> {
+        let name = format!("{}__consumer:{}", self.conn_id, name);
+        stream
+            .get_or_create_consumer(
+                name.as_str(),
+                PullConsumerConfig {
+                    durable_name: Some(name.to_owned()),
+                    ..config.unwrap_or_default()
+                },
+            )
+            .await
+            .map_err(|e| NatsError::CreateConsumerFailed {
+                name: name.clone(),
+                source: e,
+            })
+    }
+
+    pub async fn publish(
+        &self,
+        subject: Subject,
+        payload: Bytes,
+    ) -> BoxedResult<()> {
+        let subject_prefixed = subject.with_prefix(&self.conn_id);
+        let _ = self
+            .validate_payload(payload.clone(), &subject_prefixed)
+            .await;
+
+        let context = self.jetstream.as_ref();
+        let ack_future = context.publish(subject_prefixed, payload).await?;
+        ack_future.await?;
+        Ok(())
+    }
+
+    #[cfg(test)]
+    pub async fn setup_container() -> Result<
+        testcontainers::ContainerAsync<testcontainers::GenericImage>,
+        testcontainers::TestcontainersError,
+    > {
+        use testcontainers::{
+            core::{IntoContainerPort, WaitFor},
+            runners::AsyncRunner,
+            GenericImage,
+            ImageExt,
+        };
+
+        std::env::set_var("TESTCONTAINERS_COMMAND", "keep");
+
+        GenericImage::new("nats", "latest")
+            .with_exposed_port(4222.tcp())
+            .with_wait_for(WaitFor::message_on_stderr("Server is ready"))
+            .with_wait_for(WaitFor::message_on_stderr(
+                "Listening for client connections on 0.0.0.0:4222",
+            ))
+            // This is needed to avoid intermittent failures
+            .with_wait_for(WaitFor::millis(2000))
+            .with_cmd(["-m", "8222", "--name", "fuel-core-nats-server", "--js"])
+            .start()
+            .await
+    }
+
+    #[cfg(test)]
+    pub async fn connect_with_testcontainer() -> anyhow::Result<(
+        Self,
+        impl std::future::Future<Output = anyhow::Result<()>>,
+    )> {
+        let container = Self::setup_container().await?;
+        let host = container.get_host().await?;
+        let host_port = container.get_host_port_ipv4(4222).await?;
+        let url = format!("{host}:{host_port}");
+        let client = NatsClient::connect(url.as_str(), "test", None).await?;
+
+        Ok((client, async move {
+            container.stop().await?;
+            container.rm().await?;
+            Ok(())
+        }))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[tokio::test]
+    async fn create_connection() -> BoxedResult<()> {
+        let (client, cleanup) =
+            NatsClient::connect_with_testcontainer().await?;
+        assert!(client.is_connected().await?);
+        cleanup.await?;
+        Ok(())
+    }
```

Comment:
Fwiw, this doesn't test our actual connection.

Reply:
Also, I'm not sure why we need TestContainers when we already contextualized the connections with subject prefixes.